### PR TITLE
Refactor testing to leverage core chart-testing capabilities and enable multiple root level charts

### PR DESCRIPTION
**Pull request by marcofranssen**
This refactor leverages the chart-testing ci values feature (https://github.com/helm/chart-testing/blob/main/pkg/chart/chart.go#L571-L605).

Doing so we can also support multiple root level charts. 

Some of the tests should in that case move into the specific root level chart once we start to break out those charts.

### CODE REVIEW

File: charts/spire/ci/external-mysql-values.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  dataStore:
+    sql:
+      databaseType: mysql
+      databaseName: spire-server
+      username: spire
+      password: sp1ff3Test
+      host: mysql.mysql
+      port: 3306
+      options:
+      - parseTime: true
```
Comment (kfox1111) [inline charts/spire/ci/external-mysql-values.yaml:11]: We need to be careful including stuff in the chart. it gets loaded all the way into the users cluster. Maybe the .helmignore file is enough to exclude this file though.

**Comment by edwbuck** [comment]
By force pushing, I'm guessing that the testing broke, because force pushing is fundamentally different than normal git operations, in the sense that it rewrites history, a feature that generally is against the idea of source code control.

@marcofranssen Can you fix this to see all of the CI tests pass?  If not, would you close this PR after capturing the changes and submit a new one, which should unstick the CI testing?

**Comment by marcofranssen** [comment]
The broken test have nothing to do with force pushing 🤷‍♂️ . Locally I have the same bug. The problem is in the uninstall of OIDC. Just need some time to fix that bug.

**Comment by edwbuck** [comment]
@marcofranssen Thanks for the update, and the root cause.  I'll be patient, just trying to get stuff out of review.... :)

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/tests/test-connection.yaml [hunk 2]
```diff
@@ -33,7 +33,7 @@ spec:
     - name: curl-ingress
       image: {{ template "spire-lib.image" (dict "image" .Values.tests.bash.image "global" .Values.global) }}
       command: ['curl']
-      args: ['-s', '-f', 'http://{{ index .Values.config.domains 0 }}/.well-known/openid-configuration']
+      args: ['-s', '-f', 'http://{{ (index .Values.ingress.hosts 0).host }}/.well-known/openid-configuration']
```
Comment (marcofranssen) [inline charts/spire/charts/spiffe-oidc-discovery-provider/templates/tests/test-connection.yaml:36]: Prevents issues when this order is defined differently.

https://github.com/spiffe/helm-charts/pull/324/files#diff-dd133a69512e088172a3954ce5dc656ee751f7871c7d3bb839ca0f6a2675f897R29-R31

This resolved the failing test which I struggled with for a while.

**Comment by marcofranssen** [comment]
@faisal-memon if we decide to move this one forward we have to update the required ci-jobs.

My proposal is to require the `test (v1.27.2)` and the `examples`.

File: .github/kind/conf/kind-config.yaml [hunk 3]
```diff
@@ -15,3 +15,4 @@ kubeadmConfigPatches:
         # admission-control-config-file: /etc/kubernetes/pki/admctrl/admission-control.yaml
 nodes:
   - role: control-plane
+  - role: worker
```
Comment (marcofranssen) [inline .github/kind/conf/kind-config.yaml:18]: allows testing with multiple agents

**Review by kfox1111 — COMMENTED** [review]
In general, the approach seems ok to me.

A couple tests have a fair bit of regression that would be nice to fix. Some hardcoded passwords related to that are also a bit concerning.

The rest is fairly minor things. Details inline.

File: .github/tests/pre-install.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/pre-install.sh:1]: Yes we can do follow-up refactors to get rid of more plumbing etc. Did the bare minimum to get testing of multiple charts working.

File: .github/tests/pre-install.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+SCRIPT="$(readlink -f "$0")"
+SCRIPTPATH="$(dirname "${SCRIPT}")"
+DEPS="${SCRIPTPATH}/dependencies"
+
+# Set repo and version env variables
+REPOS=$(jq -r '.[] | "export " + ("HELM_REPO_" + .name | ascii_upcase | gsub("-";"_")) + "=" + .repo' "${SCRIPTPATH}/charts.json")
+VERSIONS=$(jq -r '.[] | "export " + ("VERSION_" + .name | ascii_upcase | gsub("-";"_")) + "=" + .version' "${SCRIPTPATH}/charts.json")
+eval "$REPOS"
+eval "$VERSIONS"
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/pre-install.sh:13]: wanted to make things more transparant for the end users. There is so much abstraction and plumbing going on that things aren't as straight forward as I want them to be anymore.

Maybe some follow-up refactors to use a Make native approach I can simplify things further

File: charts/spire/ci/external-mysql-values.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+spire-server:
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply charts/spire/ci/external-mysql-values.yaml:1]: The test covers still the same, but instead of generating couple of things I hardcoded them. 

1. to make it work without additional plumbing around chart-testing
2. to simplify the testcases and make it more transparant for contributors
3. the tests in CI are intended more for small unit testing, e.g. testing some non default values.

For more advanced tests we can still do the examples, but that will require releases of the charts as soon we break them out due to the dependencies we will then get on certain versions.

File: examples/external-mysql/values.yaml [hunk 7]
```diff
@@ -2,7 +2,8 @@ spire-server:
   dataStore:
     sql:
       databaseType: mysql
-      databaseName: spire
-      host: mysql
+      databaseName: spire-server
+      host: mysql.mysql
       port: 3306
       username: spire
+      password: sp1ff3Test
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply examples/external-mysql/values.yaml:9]: Yes thought about that, but also intended here to make examples more straightforward. If there is to much plumbing in examples it also complicates it for our consumers to follow the example and therefore looses its value as an example.

In the end `power comes with great responsibility` and users could copy paste things from many places on the internet or chatgpt. They do need to do their due diligence.

We could however add a note in the README.md to warn them about this hardcoded password. Let me add that at the minumum.

File: examples/external-postgresql/values.yaml [hunk 8]
```diff
@@ -2,9 +2,10 @@ spire-server:
   dataStore:
     sql:
       databaseType: postgres
-      databaseName: spire
-      host: postgresql
+      databaseName: spire-server
+      host: postgresql.postgresql
       port: 5432
       username: spire
+      password: sp1ff3Test
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply examples/external-postgresql/values.yaml:9]: https://github.com/spiffe/helm-charts/pull/324#discussion_r1239545211

File: .github/workflows/helm-chart-ci.yaml [hunk 9]
```diff
@@ -165,39 +144,79 @@ jobs:
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - name: Create kind cluster
+        uses: helm/kind-action@v1.7.0
+        # Only build a kind cluster if there are chart changes to test.
+        with:
+          version: ${{ env.KIND_VERSION }}
+          node_image: kindest/node:v1.26.4
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:217]: as this workflow uses one version I prefer to keep it here to not abstract things to much and make it less transparant. This version is used only in a single place. Once we would use it in multiple places I would agree and move it to an env var probably. For now that seems like pre optimization, causing more complexity for users to hop through another abstraction to figure things out.

The reasoning to run on 1.26.4 is because that is the current `stable` version. All the other tests run on latest greatest release.

File: Makefile [hunk 10]
```diff
@@ -31,7 +30,33 @@ clean-lingering-resources:
 		&>/dev/null || true
 
 .PHONY: test
-test: ## Run tests using Helm chart-testing (ensure to run on dedicated test cluster)
+test: install-test-dependencies ## Run tests using Helm chart-testing (ensure to run on dedicated test cluster)
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply Makefile:33]: chart-testing can not distinguish those. You either run all tests in the ci folder or you run all of them.

What we can split later though is to run them just for a single root chart, if I'm not mistaking.

E.g. some of those ci values would then move with the break out chart.

Also helm is in essence already integration testing as you always have a dependency on external system like k8s already.

File: examples/external-mysql/run-tests.sh [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+set -xe
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply examples/external-mysql/run-tests.sh:3]: I prefer flat folder structures to have everything in plain sight so users don't have to hop through folders to see the full picture. `run-tests.sh` perfectly explains the intention IMHO.

Another benefit is the complexity for CI pipeline is more straight forward without additional folders and nesting.

These are in the examples folder for a reason to show people an example how to run with external-mysql. 

Examples which we don't want to include testing in the future we simply skip the run-tests.sh

All of the refactors in this PR are done to simplify testing and make examples more transparant for people to consume.

I have added testing to the examples to cover some of the more advanced usecases which are harder to tackle with `ci/*-values.yaml`.

File: Makefile [hunk 12]
```diff
@@ -31,7 +30,33 @@ clean-lingering-resources:
...
 	@echo Running tests…
-	@ct install --debug \
-		--charts charts/spire
+	@ct install --config ct.yaml
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply Makefile]: Done.

File: .github/tests/dependencies/postgresql.yaml [hunk 13]
```diff
@@ -6,3 +6,8 @@ primary:
       drop: [ALL]
     seccompProfile:
       type: RuntimeDefault
+auth:
+  database: spire-server
+  username: spire
+  password: sp1ff3Test
+  postgresPassword: sp1ff3TestPassword
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/dependencies/postgresql.yaml:13]: Did this because it would complicate the test setup. This is reused in both examples and ci/*-values.yaml tests.

Making this on variable generated password would require again whole bunch of plumbing on the chart-testing tool.

File: .github/workflows/helm-chart-ci.yaml [hunk 14]
```diff
@@ -165,39 +144,79 @@ jobs:
...
-              --namespace "${scenario}" \
-              --target-branch ${{ github.base_ref }} \
-              --exclude-deprecated \
-              --skip-clean-up \
-              ${{ (matrix.values != 'default' && '--helm-extra-set-args "--values=${TEST_DIR}/values.yaml ${EXTRA_HELM_ARGS}"') || '' }}
-          fi
-        env:
-          TEST_DIR: .github/tests/${{ matrix.values }}
+          ct install --config ct.yaml \
+            --target-branch ${{ github.base_ref }}
```
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:159]: No this installs every single `ci/*-values.yaml`.

The chart-testing tool will loop all those *-values.yaml files one by one runnning

1. helm install
2. helm test
3. kubectl get events
4. kubectl get logs
5. helm uninstall

for each of those values files. and for each of the charts. There are ways to run those only for the charts that have changes. We stripped that a while ago, but might be useful to reintroduce once we have multiple root charts.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply Makefile]: its not chart-testing that will use the other target. I think if we want to test a lot of the options, we can't stick to just integration testing. we will need unit testing like as proposed here: https://github.com/spiffe/helm-charts/pull/260/files#diff-a86124ab60e89d8dbec3b476b5a21077f58272f61f2007b0136c84e609738bb2

We can then do a lot of lighter weight tests to ensure config does get passed through properly, without having to deploy whole integration instance tests for every option.

File: .github/tests/dependencies/postgresql.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply .github/tests/dependencies/postgresql.yaml:13]: It is a regression of the tests though. So we should make sure we retain the same functionality....

We could do that part of the tests using unit testing instead, to ensure that passwords are passed through to the generated config. Would that work? We can adapt https://github.com/spiffe/helm-charts/pull/260/files#diff-a86124ab60e89d8dbec3b476b5a21077f58272f61f2007b0136c84e609738bb2 to do that kind of test, then I'm ok with doing a hardcoded pw test in the integration test.

File: .github/tests/pre-install.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply .github/tests/pre-install.sh:1]: ok.

File: .github/tests/pre-install.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply .github/tests/pre-install.sh:13]: I don't see how it is difficult on users for them to see it source the other file. A comment on what it does when sourcing is probably sufficient? (ex: #include common environment variables for chart repos and versions) It may even make it more readable that way. It also makes it harder on us if we want to update the logic later as we cant change it in one place anymore, and when contributors want to make new ones, they have to copy it yet again.

File: .github/workflows/helm-chart-ci.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml:217]: It would be just to keep all the versions together in the config. It would be cleaner that way imo. But we can skip it too. Its not that big a deal.

File: .github/workflows/helm-chart-ci.yaml [hunk 15]
```diff
@@ -165,39 +144,79 @@ jobs:
...
+        uses: helm/kind-action@v1.7.0
+        # Only build a kind cluster if there are chart changes to test.
+        with:
+          version: ${{ env.KIND_VERSION }}
+          node_image: kindest/node:v1.26.4
+          config: .github/kind/conf/kind-config.yaml
+          verbosity: 1
+
+      - name: Install and test example
+        run: ${{ matrix.example }}/run-tests.sh
```
Comment (kfox1111) [inline .github/workflows/helm-chart-ci.yaml:222]: Looking at the pr action run list, I don't see this run per the versions we support. Something missing iterator thing here?

File: charts/spire/ci/external-mysql-values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/ci/external-mysql-values.yaml:1]: > The test covers still the same, but instead of generating couple of things I hardcoded them.
> 
> 1. to make it work without additional plumbing around chart-testing
> 2. to simplify the testcases and make it more transparant for contributors
> 3. the tests in CI are intended more for small unit testing, e.g. testing some non default values.

Its not unit testing though, its still integration testing. And its not testing as much as before. We need to come up with a way to get those tests back in. A unit testing framework to do unit tests would be an ok way to solve it I think.

> For more advanced tests we can still do the examples, but that will require releases of the charts as soon we break them out due to the dependencies we will then get on certain versions.

File: examples/external-mysql/values.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply examples/external-mysql/values.yaml:9]: It would be possible for the run-tests.sh to include the password hardcoded via --set and have the values file have it commented out with a value of CHANGEME, indicating to the user they should set it to something if they are to deploy it themselves using the values file? Maybe a good middle ground.

[the remark this answers is not in the document]
Comment (marcofranssen) [reply examples/external-mysql/values.yaml]: Smart idea. Done.

File: examples/external-postgresql/values.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply examples/external-postgresql/values.yaml]: See previous. Handled that suggestion here as well.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply Makefile]: Whooaaa, for me that really feels like going a bridge too far with testing. I haven't seen any chart that is even testing on the level we already do. Most charts keep it even more to the basics not even testing with value overrides. I don't want to end up in the situation we have to maintain a whole bunch of self invented test approaches.

Are we trying to resolve how to develop and test Helm charts? Or are we trying to get a Helm chart for Spire.

IMHO such a unittesting thing needs an additional tool like chart-testing to make that usable without a lot of plumbing here.

File: .github/tests/dependencies/postgresql.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/dependencies/postgresql.yaml:13]: What functionality are we not testing anymore? I only moved them? tests are still the same.

File: .github/workflows/helm-chart-ci.yaml (hunk 15 above)
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:222]: Intentionally only test the examples on the latest version, all the ci/*-values.yaml are tested on all versions.

Examples are essentially a combination of what is already tested there, so therefore saving some compute here on these tests.

Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml:222]: I would think the examples are the ones we most want to ensure work on any cluster the user might be deploying on, so should be well tested on various versions too.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply Makefile]: Its a much lighter weight way of testing so that the heavy weight things such as the integration tests don't need to be near as through. The go unit test stuff is basically one file with a few lines per test. Not particularly onerous.

File: .github/tests/dependencies/postgresql.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/dependencies/postgresql.yaml:13]: The hardcoded value proofs the same as it wouldn't work without defining those values either. No need to make things more complex then necessary.

File: .github/workflows/helm-chart-ci.yaml (hunk 15 above)
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:222]: The examples are just a combination of the ci tests. Those already cover the same ground and would already cause the failure in case the other k8s version wouldn't support things. It would just be a waste of resources to run the examples again on 3 k8s versions.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply Makefile]: Revamped.

File: charts/spire/ci/external-mysql-values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply charts/spire/ci/external-mysql-values.yaml:1]: Sure, but there haven't been tests removed here which was you original comment. We are overriding the chart default with a fixed value as opposed to a generated value. That doesn't test less neither more. So imho we didn't drop any testing here.

File: .github/workflows/helm-chart-ci.yaml (hunk 15 above)
Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml:222]: I don't think thats true... take for example the production example. If we make a change to it, and we're only testing on the newest, it may break for n-1 and we wouldn't know it. I don't see any of the ct test cases with the same functionality, so its not just a combination of existing ct tested features.

File: .github/tests/pre-install.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/tests/pre-install.sh]: Moved back to central script for now. Till we further refactor and simplify in follow-up PRs.

**Comment by marcofranssen** [comment]
@kfox1111 handled the remaining feedback.

- [x] Renamed Makefile task for testing to allow adding more test types in future.
- [x] Removed the duplication on parsing the helm charts versions by reintroducing a script.
- [ ] Didn't update matrix to run all examples again on all 3 k8s versions. In essence all is already individual features are already covered in the chart tests on each k8s version. Running the examples (combinations of those individual tests) again on all 3 versions of k8s would be a waste of resources. k8s specific failures would already be captured in the chart tests themself and therefore already fail the build on that particular version.

File: .github/workflows/helm-chart-ci.yaml (hunk 15 above)
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:222]: Which exact ones do you mis?

**Comment by kfox1111** [comment]
Still cant see whats happening with the tests: https://github.com/spiffe/helm-charts/actions/runs/5448368164/jobs/9911467882?pr=324. Its seems to be skipping the runs. Could you update things to always run again like it used to please?

Also, it seems to be waiting on 4 tests that may not ever run? Know whats going on there?

**Comment by kfox1111** [comment]
Maybe related to the funniness with the 4 blocked tests, the pr page is showing the results for
https://github.com/spiffe/helm-charts/actions/runs/5448368164
but https://github.com/spiffe/helm-charts/actions/runs/5448368165 is a newer run that seems to be running all the tests ok.

I do see the normal tests executing out of ci dir now. so that does seem to work ok.

But there is a drawback to this approach I hadn't noticed before.  Its running all those tests serially per k8s version. so its taking roughly 12 minutes wall clock for the ct tests to run. Much longer feedback times on pr's with this approach then the previous one.

Could we break those out with a matrix job too so they can run in parallel on GH?

**Comment by marcofranssen** [comment]
> Maybe related to the funniness with the 4 blocked tests, the pr page is showing the results for [spiffe/helm-charts/actions/runs/5448368164](https://github.com/spiffe/helm-charts/actions/runs/5448368164) but [spiffe/helm-charts/actions/runs/5448368165](https://github.com/spiffe/helm-charts/actions/runs/5448368165) is a newer run that seems to be running all the tests ok.
> 
> I do see the normal tests executing out of ci dir now. so that does seem to work ok.
> 
> But there is a drawback to this approach I hadn't noticed before. Its running all those tests serially per k8s version. so its taking roughly 12 minutes wall clock for the ct tests to run. Much longer feedback times on pr's with this approach then the previous one.
> 
> Could we break those out with a matrix job too so they can run in parallel on GH?

No we can't break out the tests into separate matrix. chart-testing framework runs all those tests sequentially including `uninstall`, something we didn't test before. If we want to break out the charts we need this. To reduce the test times chart-testing can only run the tests for the charts that have changed in a PR, so once we break out the charts, some tests will move with that chart. If that chart isn't touched in a certain PR we can configure it to be skipped so it runs only the necessary tests there, while we still run the examples for everything.

**Comment by marcofranssen** [comment]
> Still cant see whats happening with the tests: [spiffe/helm-charts/actions/runs/5448368164/jobs/9911467882?pr=324](https://github.com/spiffe/helm-charts/actions/runs/5448368164/jobs/9911467882?pr=324). Its seems to be skipping the runs. Could you update things to always run again like it used to please?
> 
> Also, it seems to be waiting on 4 tests that may not ever run? Know whats going on there?

This is the renaming/restructuring of jobs. We have to change the branch protections to match the new names, but can only do that at merge time to not block other PRs.

File: .github/workflows/helm-chart-ci.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply .github/workflows/helm-chart-ci.yaml:156]: They only work for the examples. chart-testing also performs uninstall and does those checks before running a uninstall. We never tested the uninstall. Due to uninstall we can't do those checks again ourself after ct finishes.

I did open an issue to build support into ct to print that output to GITHUB_STEP_SUMMARY (if I find time I will implement it in the ct binary myself). 

https://github.com/helm/chart-testing/issues/558

**Review by faisal-memon — APPROVED** [review]

**Comment by edwbuck** [comment]
If this change goes forward, I will lose the ability to migrate my customers onto this project.  I've mentioned before, they see Helm as a package manager of an Application, and minimally they see the "Server, Agent, Container Manager / Registrar, and SPIFFE CSI Drivers" as part of the core application.  They don't mind seeing Tornjak as a sub-chart, and they are considering possibly using sub-charts for their database deployments (which are currently heavily managed by terraform as they are Athena databases). 

We should not make the server an Application, and the agent a different application and then require that they both be tied together by a sub-chart style deployment.  I would ask that we reconsider moving the server and the agent, minimally, together into the root chart.

**Comment by kfox1111** [comment]
Discussion of something that affects this PR started here:
https://spiffe.slack.com/archives/C04BURZP1KK/p1688575164862469

**Comment by kfox1111** [comment]
> If this change goes forward, I will lose the ability to migrate my customers onto this project. I've mentioned before, they see Helm as a package manager of an Application, and minimally they see the "Server, Agent, Container Manager / Registrar, and SPIFFE CSI Drivers" as part of the core application. They don't mind seeing Tornjak as a sub-chart, and they are considering possibly using sub-charts for their database deployments (which are currently heavily managed by terraform as they are Athena databases).
> 
> We should not make the server an Application, and the agent a different application and then require that they both be tied together by a sub-chart style deployment. I would ask that we reconsider moving the server and the agent, minimally, together into the root chart.

This style of application breakup within a chart is not uncommon. For example, gitlab on prem is often used by enterprise customers and is architected with subcharts:
https://gitlab.com/gitlab-org/charts/gitlab/-/tree/master/charts
Same with kube-prometheus-stack:
https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/Chart.yaml#L41-L57

Is the pushback related to a feeling that it shouldn't be done that way, or is there evidence that harm is done when structured that way? Would it help to get feedback from the helm dev team itself, speaking to the legitimacy or illegitimacy of using nested charts for this sort of application?

**Comment by faisal-memon** [comment]
> If this change goes forward, I will lose the ability to migrate my customers onto this project. I've mentioned before, they see Helm as a package manager of an Application, and minimally they see the "Server, Agent, Container Manager / Registrar, and SPIFFE CSI Drivers" as part of the core application. They don't mind seeing Tornjak as a sub-chart, and they are considering possibly using sub-charts for their database deployments (which are currently heavily managed by terraform as they are Athena databases).
> 
> We should not make the server an Application, and the agent a different application and then require that they both be tied together by a sub-chart style deployment. I would ask that we reconsider moving the server and the agent, minimally, together into the root chart.

@edwbuck We will still support the aggregated chart, just provide options for people to deploy components separately for advanced deployments. Functionally it will still be the same, implementation details would be different. Anything specific that would prevent adoption by your customers?

**Comment by edwbuck** [comment]
> @edwbuck We will still support the aggregated chart, just provide options for people to 
> deploy components separately for advanced deployments. Functionally it will still be the 
> same, implementation details would be different. Anything specific that would prevent 
> adoption by your customers?

Basically they see a cluster as an Application.  The only exception to this is their root clusters (they run on nested spire) where the root cluster is also an Application, but the child clusters are again single Applications with an additional (optional) Agent that is part of the root cluster.  For their standalone clusters, they use the child cluster with the "upstream" agent turned off.  

They rely heavily on simple / easy to interpret templates, and they are not keen on having to dig into one Chart's file to ensure it coordinates with another Chart's file, but they are more inclined to grep across multiple .tpl files in the same Chart.  

Part of that is because they also have monitoring and management through other tooling that would over-report each component as a different application, preventing them from getting the single view of the entire cluster.  That also impacts their DataDog collection system, as the multiple deployments for a single conceptual cluster would affect their auto-labeling system such that the metrics would be more disjoint.

In short, my requests to at least have the core components of SPIRE be delivered in one Helm Chart (the extras can easily be sub-charted, they'll likely deploy them independently anyway) are not just interesting design ideas, it's what they want.

Currently, they are committed to maintaining their own charts, and their commentary on this project isn't worth repeating.  Let's just say they're not fans.

**Comment by marcofranssen** [comment]
> > @edwbuck We will still support the aggregated chart, just provide options for people to
> > deploy components separately for advanced deployments. Functionally it will still be the
> > same, implementation details would be different. Anything specific that would prevent
> > adoption by your customers?
> 
> Basically they see a cluster as an Application. The only exception to this is their root clusters (they run on nested spire) where the root cluster is also an Application, but the child clusters are again single Applications with an additional (optional) Agent that is part of the root cluster. For their standalone clusters, they use the child cluster with the "upstream" agent turned off.
> 
> They rely heavily on simple / easy to interpret templates, and they are not keen on having to dig into one Chart's file to ensure it coordinates with another Chart's file, but they are more inclined to grep across multiple .tpl files in the same Chart.
> 
> Part of that is because they also have monitoring and management through other tooling that would over-report each component as a different application, preventing them from getting the single view of the entire cluster. That also impacts their DataDog collection system, as the multiple deployments for a single conceptual cluster would affect their auto-labeling system such that the metrics would be more disjoint.
> 
> In short, my requests to at least have the core components of SPIRE be delivered in one Helm Chart (the extras can easily be sub-charted, they'll likely deploy them independently anyway) are not just interesting design ideas, it's what they want.
> 
> Currently, they are committed to maintaining their own charts, and their commentary on this project isn't worth repeating. Let's just say they're not fans.

This PR is not even breaking out charts at this stage. This chart is all about using the industry standard tooling instead of all our custom scripting. Lets have that discussion of breaking out charts somewhere else. Even if we are not breaking out charts this refactor cleans up already +-140 lines of custom plumbing code.

**Comment by marcofranssen** [comment]
> If this change goes forward, I will lose the ability to migrate my customers onto this project. I've mentioned before, they see Helm as a package manager of an Application, and minimally they see the "Server, Agent, Container Manager / Registrar, and SPIFFE CSI Drivers" as part of the core application. They don't mind seeing Tornjak as a sub-chart, and they are considering possibly using sub-charts for their database deployments (which are currently heavily managed by terraform as they are Athena databases).
> 
> We should not make the server an Application, and the agent a different application and then require that they both be tied together by a sub-chart style deployment. I would ask that we reconsider moving the server and the agent, minimally, together into the root chart.

So those same customers of yours, do they demand the same from any opensource project. Telling them on how to organize their Go packages, and if such a project to decide to move things into a separate library etc? Maybe we should ask the spire project to get rid of the spiffe module and put that in the same spire repo as well, because that would for me also be easier to find the code as opposed to pulling in a dependency managed in another repo 🤷‍♂️ .

What we are talking about here is internals. The public API for them will still be the same. They still install the spire Helm chart. They still provide the same values. We can't have others decide how we organize things internally for other usecases. So in terms of golang we are talking about things like. Put all in the same package vs having multiple packages vs having external dependencies.

**Comment by edwbuck** [comment]
> > If this change goes forward, I will lose the ability to migrate my customers onto this project. I've mentioned before, they see Helm as a package manager of an Application, and minimally they see the "Server, Agent, Container Manager / Registrar, and SPIFFE CSI Drivers" as part of the core application. They don't mind seeing Tornjak as a sub-chart, and they are considering possibly using sub-charts for their database deployments (which are currently heavily managed by terraform as they are Athena databases).
> > We should not make the server an Application, and the agent a different application and then require that they both be tied together by a sub-chart style deployment. I would ask that we reconsider moving the server and the agent, minimally, together into the root chart.
> 
> So those same customers of yours, do they demand the same from any opensource project. Telling them on how to organize their Go packages, and if such a project to decide to move things into a separate library etc? Maybe we should ask the spire project to get rid of the spiffe module and put that in the same spire repo as well, because that would for me also be easier to find the code as opposed to pulling in a dependency managed in another repo 🤷‍♂️ .
> 
> What we are talking about here is internals. The public API for them will still be the same. They still install the spire Helm chart. They still provide the same values. We can't have others decide how we organize things internally for other usecases. So in terms of golang we are talking about things like. Put all in the same package vs having multiple packages vs having external dependencies.

I will reach back to the customer with this messaging.  Thank you.  While the two charts appear very different, which is stalling them from adopting, if the differences are truly internal, this will help with future migration efforts.

**Review by kfox1111 — APPROVED** [review]
I still have concerns its taking so long to do runs, but since we think we can get it reduced in follow on patches, I'll go along with that.

The rest looks ok.

**Comment by kfox1111** [comment]
Fixed the merge conflict
